libraries/exoplayer/src/main/java/androidx/media3/exoplayer/drm/DefaultDrmSession.java

```diff
@@ -666,6 +672,12 @@ public void handleMessage(Message msg) {
             break;
           case MSG_KEYS:
             response = callback.executeKeyRequest(uuid, (KeyRequest) requestTask.request);
+            if (currentKeyRequestInfo != null) {
+              LoadEventInfo loadEventInfo = callback.getLastLoadEventInfo();
```

Comment:
> Unfortunately because of the threading model used here, I don't think we can **always** guarantee that `getLastLoadEventInfo()` here corresponds to the request made above on L674.
> 
> This can happen because each `DefaultDrmSession` instance has its own request thread, but two instances can easily share the same `MediaDrmCallback` instance.

Ah, great catch...  I feel pretty stupid, given that  I just solved this same problem with my custom `LoadErrorHandlingPolicy` I wound up keeping the state object key'd by the load `taskId`

When I get back from vacation next week I'll move my original commits to a branch and download your changes (the force push would make a mess if I just pulled it) so I can see what is going on and have a closer look at your suggestions.   Thanks!

Reply:
> the force push would make a mess if I just pulled it

Ah sorry, the force push was part of me planning to send this PR for review internally before I realised this larger issue. I didn't push any 'meaningful' changes though, just formatted the changes in a commit, then rebased on top of the latest `main` commit. You should be able to force the branch pointer back to the 'old' version of https://github.com/androidx/media/commit/ebec29367295da0ba7ba8d20c27dc1e1a1383950 which I think is https://github.com/androidx/media/commit/9d27d45a65, then do your own force push to 'take back control' of the PR.

Probably something like this (though I'm no git expert, so there may be a better way to achieve this!):

```shell
$ git branch --force p-add-loadinfo-to-ondrmkeysloaded 9d27d45a65
$ git push --force
```

Reply:
> Unfortunately because of the threading model used here, I don't think we can **always** guarantee that `getLastLoadEventInfo()` here corresponds to the request made above on L674.
> 
> This can happen because each `DefaultDrmSession` instance has its own request thread, but two instances can easily share the same `MediaDrmCallback` instance.

Yes, stashing the `LoadEventInfo` anywhere other than the `RequestTask`, which is new'd for each request will violate the statelessness of whatever it is stuffed into.   There is no clean way out of just doing this.

> To resolve this I think its best to tightly couple each request to the reported `LoadEventInfo`, so we can keep `MediaDrmCallback` impls 'stateless' so it doesn't matter if they're invoked on multiple threads at the same time.
> 
> Which I think leads us to either:
> 
> 1. Pass `TransferListener` into `MediaDrmCallback.executeXXXRequest` (my previous suggestion in [Adds load info to `DrmSessionEventListeneronDrmKeysLoaded()` #1134 (comment)](https://github.com/androidx/media/pull/1134#discussion_r1503874316))
> 2. Enrich the return type of `MediaDrmCallback.executeKeyRequest` to `KeyResponse`, which has `byte[] data` field and optional `@Nullable LoadEventInfo` too.
>    

also, we could

3. Add to or lift a common base class from `ProvisionRequest` and `KeyRequest` that includes the `LoadEventInfo` as a member.

Number 3 avoids a signature change for `executeXXXRequest`,  but it seems uglier than your number 2, which is pretty strait forward.

So I would vote for your number 2.    I can do it or feel free to branch the current code and make a pull request onto my repo with your idea, I can build and test it.   Branches and pull request are best way to collaborate (once we agree the basic approach),  but no worries about the force push and thanks for re-basing to keep up with head of `main` (this is a big set of changes to keep on the side for long)

Thanks again for catching this.

Reply:
@icbaker  finished the rebase to handle the move of modified code `DrmUtil.executePost()`.   In the process I addressed the thread safety of the `LoadEventInfo` object.  This is now entirely encapsulated in the `MediaDrmCallback.executeXXXRequest` methods, they now return an object with the response data and the `LoadEventInfo`

All of the test cases have been modified to create some manner of `LoadEventInfo` as well.

If this looks good we can stop here (this code is difficult to rebase) and leave the tech-debt to harden the test cases a bit to actually check for the `LoadEventInfo` (perhaps at least the `DefaultAnalyticsCollectorTest`, that would be a pretty quick add)

Let me know.